### ISSUE TITLE
The debug build process is slow.

### ISSUE THREAD

**Issue by harryfei** [Wilfred/remacs#823]
Everytime I change some code in rust, then I rebuild the project, it takes more than 5 minutes to done.  It's so slow. I don't know why. It makes our development workflow less efficient.

**Comment by Wilfred** [comment]
I believe this is due to bindgen. I've found `cargo check` to be quick, and that's how I usually work. Still, perhaps there's scope for making rebuilds faster.

**Comment by steinuil** [comment]
From my printf-debugging tests it seems to be bindgen's fault.

**Comment by db48x** [comment]
Yes, running bindgen takes a while. We also run it too often; we don't need to run it when a Rust source file changes, only when a C header file changes. However, our build.rs script does things that do need to be rerun when a Rust file changes, so we're currently forced to run the whole thing any time any source file changes. When I added bindgen I took a little bit of time to try to narrow down how often we ran it, but I stopped short of completely reinventing the whole dependency tracking mechanism provided by a simple Makefile. Maybe there's some convenient crate we could use, or maybe we can convince cargo to let us have multiple build scripts.

**Comment by steinuil** [comment]
Would simply using a makefile and a separate build script for bindgen work for the moment?

**Comment by db48x** [comment]
That's an idea. We could completely extract the bindgen part of build.rs into a separate program which is compiled first, and then run when needed.

**Comment by carlosmn** [comment]
I'm attempting to do this extraction in #1185 which should help with only rebuilding when the C headers have changed.

I've also discovered that bindgen supports printing out its phases itself so here's the ones I got just now

```
  time:  5915.497 ms.   parse
  time:     0.001 ms.   process_replacements
  time:    95.923 ms.   deanonymize_fields
  time: 33250.418 ms.   compute_whitelisted_and_codegen_items
  time:  4208.238 ms.   compute_has_vtable
  time: 23377.932 ms.   compute_sizedness
  time:  4245.175 ms.   compute_has_destructor
  time: 20197.130 ms.   find_used_template_parameters
  time: 17804.191 ms.   compute_cannot_derive_debug
  time: 25167.290 ms.   compute_cannot_derive_default
  time: 26895.908 ms.   compute_cannot_derive_copy
  time:  4335.181 ms.   compute_has_type_param_in_array
  time:     0.001 ms.   compute_has_float
  time:     0.000 ms.   compute_cannot_derive_hash
  time:     0.001 ms.   compute_cannot_derive_partialord_partialeq_or_eq
  time: 34294.196 ms.   codegen
  time:  2597.379 ms.   rustfmt_generated_string
```

We seem to spend ~24s out of ~3.3m between `compute_has_destructor` and `find_used_template_parameters` which should be unnecessary in C code.

While this is just speculation for now, we can then also see that `compute_cannot_derive_debug`, `compute_cannot_derive_default` and `compute_cannot_derive_copy` take a suspiciously similar amount of time where it looks like it might be going over the list of symbols (or whatever) once in each check instead of iterating once and calculating whatever it needs to calculate.
